consensus/istanbul/core/core.go

```diff
@@ -769,13 +772,14 @@ func (c *core) verifyProposal(proposal istanbul.Proposal) (time.Duration, error)
 	} else {
 		logger.Trace("verification status cache miss")
 
-		duration, err := c.backend.Verify(proposal)
+		result, duration, err := c.backend.Verify(proposal)
 		logger.Trace("proposal verify return values", "duration", duration, "err", err)
 
 		// Don't cache the verification status if it's a future block
 		if err != consensus.ErrFutureBlock {
 			c.current.SetProposalVerificationStatus(proposal.Hash(), err)
 		}
+		c.current.SetStateProcessResult(proposal.Hash(), result)
```

Comment:
If this is a future block, setting the result in the cache here causes an inconsistent state because the block evaluated in the future will achieve a result that does not match what is in the cache (i.e. it will result a full state, but the cache holds `nil`). I would recommend either writing to the cache only if the error is non-nill, or only if it is not `consensus.ErrFutureBlock`

Reply:
+1

Reply:
I handled it here: https://github.com/celo-org/celo-blockchain/blob/d1d3805c15fdc754a8c948f3fcf345eac30ad01e/consensus/istanbul/backend/backend.go#L497
we get a non-nil result from the cache only when error is nil.